libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,115 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+#include <vector>
+#include <variant>
+
+namespace solidity::frontend
+{
+
+/* Creates a Function call graph for a contract.
+ * Includes the following special nodes:
+ *  - CreationRoot: All calls made at contract creation originate from this node
+ *  - CreationDispatch: Represents the creation dispatch function
+ *  - ExternalDispatch: Represents the runtime dispatch for all external functions
+ *  - InternalDispatch: Represents the runtime dispatch for internal function pointers and complex expressions
+ *
+ *  Nodes are a variant of either the enum SpecialNode or an ASTNode pointer.
+ *  ASTNodes are usually inherited from CallableDeclarations
+ *  (FunctionDefinition, ModifierDefinition, EventDefinition) but for functions
+ *  without declaration it is directly the FunctionCall AST node.
+ *
+ *  Functions that are not called right away as well as functions without
+ *  declarations have an edge to the dispatch node. */
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	enum class SpecialNode { Unset, CreationRoot, CreationDispatch, InternalDispatch, ExternalDispatch };
+
+	typedef std::variant<ASTNode const*, SpecialNode> Node;
+
+	struct CompareByID
+	{
+		using is_transparent = void;
+
+		bool operator()(Node const& _lhs, Node const& _rhs) const
+		{
+			if (_lhs.index() != _rhs.index())
```

Comment:
Most of this is already done by the comparison operator for the variant - can we somehow reuse that by only "redefining" comparison on ASTNode pointers?

Reply:
I am not really sure how..

Reply:
ok!